src/Diagrams/Backend/SVG.hs

```diff
@@ -259,15 +262,55 @@ instance SVGFloat n => Backend SVG V2 n where
 
   adjustDia c opts d = adjustDia2D sizeSpec c opts (d # reflectY)
 
+compressStyles :: SVGFloat n => RTree SVG V2 n Annotation -> RTree SVG V2 n Annotation
+compressStyles (Node n []) = Node n []
+compressStyles (Node n rs) = Node n (map compressStyles $ xs <> other)
+  where
+    xs = concatMap compress allSty
+    compress (Node (RStyle s) ss) =
+      map (\(Node (RStyle s') xs) -> Node (RStyle (s' <> s)) xs) ss
+    compress _ = error "Style nodes only"
+    (allSty, other) = partition styleChildren rs
+    styleChildren (Node (RStyle _) cs) = all isStyle cs
+    styleChildren _                    = False
+    isStyle (Node m _) = case m of {RStyle _ -> True; _ -> False}
+
 rtree :: SVGFloat n => RTree SVG V2 n Annotation -> Render SVG V2 n
 rtree (Node n rs) = case n of
-  RPrim p                 -> render SVG p
-  RStyle sty              -> R $ local (over style (<> sty)) r
+  RPrim p -> render SVG p
+  RStyle sty              ->
+    let r' = local (over style (<> sty)) r
+    in R $ stylize sty r'
+  -- RStyle sty ->
+  --   let (prims, others) = partition isPrim rs
+  --   in mappend (R $ stylize sty (unR $ foldMap rtree prims))
+  --              (R $ local (over style (<> sty)) (unR $ foldMap rtree others))
   RAnnot (OpacityGroup o) -> R $ g_ [opacity_ $ toText o] <$> r
-  RAnnot (Href uri)       -> R $ a_ [xlinkHref_ $ T.pack uri] <$> r
-  _                       -> R r
+  RAnnot (Href uri) -> R $ a_ [xlinkHref_ $ T.pack uri] <$> r
+  _ -> R r
   where
     R r = foldMap rtree rs
+    isPrim (Node (RPrim _) _) = True
+    isPrim (Node _ _)         = False
+
+stylize :: SVGFloat n => Style V2 n -> SvgRenderM n  -> SvgRenderM n
+stylize sty svg = do
+  SvgRenderState _idClip idFill idLine <- get
+  Environment sty preT <- ask
```

Comment:
But this style is the accumulated style, you only need the style from the node (which is the `sty` in the function argument). Is this why you're not seeing any reductions in size?